src/Compilers/CSharp/Portable/Symbols/TypeSymbolWithAnnotations.cs

```diff
@@ -515,37 +515,83 @@ public bool IsAtLeastAsVisibleAs(Symbol sym, ref HashSet<DiagnosticInfo> useSite
         internal TypeSymbolWithAnnotations SubstituteTypeCore(AbstractTypeMap typeMap, bool withTupleUnification)
         {
             var newCustomModifiers = typeMap.SubstituteCustomModifiers(this.CustomModifiers);
-            var newTypeWithModifiers = typeMap.SubstituteType(this.TypeSymbol, withTupleUnification);
-            bool newIsAnnotated = this.IsAnnotated || newTypeWithModifiers.IsAnnotated;
+            TypeSymbol typeSymbol = this.TypeSymbol;
+            var newTypeWithModifiers = typeMap.SubstituteType(typeSymbol, withTupleUnification);
+
+            if (!typeSymbol.IsTypeParameter())
+            {
+                Debug.Assert(newTypeWithModifiers.IsAnnotated == false || typeSymbol.IsNullableType());
+                Debug.Assert(newTypeWithModifiers.NonNullTypesContext.NonNullTypes == null);
+                Debug.Assert(newTypeWithModifiers.CustomModifiers.IsEmpty);
+
+                if (typeSymbol.Equals(newTypeWithModifiers.TypeSymbol, TypeCompareKind.CompareNullableModifiersForReferenceTypes) &&
+                    newCustomModifiers == CustomModifiers)
+                {
+                    return this; // substitution had no effect on the type or modifiers
+                }
+                else if (IsAnnotated == false &&
+                    NonNullTypesContext.NonNullTypes == null &&
+                    newCustomModifiers.IsEmpty)
+                {
+                    return newTypeWithModifiers;
+                }
 
-            // https://github.com/dotnet/roslyn/issues/30052: Can we use Equals instead?
-            if (this.TypeSymbolEquals(newTypeWithModifiers, TypeCompareKind.CompareNullableModifiersForReferenceTypes) &&
-                newTypeWithModifiers.CustomModifiers.IsEmpty &&
-                newIsAnnotated == this.IsAnnotated &&
-                newCustomModifiers == this.CustomModifiers)
+                return Create(NonNullTypesContext, newTypeWithModifiers.TypeSymbol, IsAnnotated, newCustomModifiers);
+            }
+
+            if (newTypeWithModifiers.Is((TypeParameterSymbol)typeSymbol) &&
+                newCustomModifiers == CustomModifiers)
             {
-                // https://github.com/dotnet/roslyn/issues/30052: We're dropping newTypeWithModifiers.NonNullTypes!
                 return this; // substitution had no effect on the type or modifiers
             }
+            else if(Is((TypeParameterSymbol)typeSymbol))
+            {
+                return newTypeWithModifiers;
+            }
+
+            bool newIsAnnotated = this.IsAnnotated || newTypeWithModifiers.IsAnnotated;
+            INonNullTypesContext newContext;
 
-            bool newIsNullableType = newTypeWithModifiers.TypeSymbol.IsNullableType();
-            if (newIsNullableType)
+            if (!newIsAnnotated)
```

Comment:
>if (!newIsAnnotated) [](start = 12, length = 20)

Minor point: Consider using `if (newIsAnnotated)` and `else if` for the remaining cases.